tests/components/enphase_envoy/test_init.py

```diff
@@ -210,12 +212,122 @@ async def test_coordinator_token_refresh_error(
             CONF_TOKEN: token,
         },
     )
-    # token refresh without username and password specified in
-    # EnvoyTokenAuthwill force token refresh error
+    # override fresh token in conftest mock_envoy.auth
     mock_envoy.auth = EnvoyTokenAuth("127.0.0.1", token=token, envoy_serial="1234")
-    await setup_integration(hass, entry)
+    # force token refresh to fail.
+    with patch(
+        "pyenphase.auth.EnvoyTokenAuth._obtain_token",
+        side_effect=EnvoyError,
+    ):
+        await setup_integration(hass, entry)
+
     await hass.async_block_till_done(wait_background_tasks=True)
     assert entry.state is ConfigEntryState.LOADED
 
     assert (entity_state := hass.states.get("sensor.inverter_1"))
     assert entity_state.state == "1"
+
+
+async def test_config_no_unique_id(
+    hass: HomeAssistant,
+    mock_envoy: AsyncMock,
+) -> None:
+    """Test enphase_envoy init if config entry has no unieuqe id."""
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        entry_id="45a36e55aaddb2007c5f6602e0c38e72",
+        title="Envoy 1234",
+        unique_id=None,
+        data={
+            CONF_HOST: "1.1.1.1",
+            CONF_NAME: "Envoy 1234",
+            CONF_USERNAME: "test-username",
+            CONF_PASSWORD: "test-password",
+        },
+    )
+    await setup_integration(hass, entry)
+    assert entry.state is ConfigEntryState.LOADED
+    assert entry.unique_id == mock_envoy.serial_number
+
+
+async def test_config_different_unique_id(
+    hass: HomeAssistant,
+    mock_envoy: AsyncMock,
+) -> None:
+    """Test enphase_envoy init if config entry has different unique id."""
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        entry_id="45a36e55aaddb2007c5f6602e0c38e72",
+        title="Envoy 1234",
+        unique_id=4321,
+        data={
+            CONF_HOST: "1.1.1.1",
+            CONF_NAME: "Envoy 1234",
+            CONF_USERNAME: "test-username",
+            CONF_PASSWORD: "test-password",
+        },
+    )
+    await setup_integration(hass, entry)
+    assert entry.state is ConfigEntryState.SETUP_RETRY
```

Comment:
It's not for this PR, but shouldn't that raise an error instead?

Reply:
> It's not for this PR, but shouldn't that raise an error instead?

Yes, it raises ConfigEntryNotReady, resulting in the ConfigEntryState.SETUP_RETRY state which is tested.

https://github.com/home-assistant/core/blob/2b3a6e5361db71cff3e44ce0c957f6951515a103/homeassistant/components/enphase_envoy/__init__.py#L28-L37

Reply:
No I mean, ConfigEntryError, since we connect to the right IP, but the serial changed, which means the current config entry can't be used with that inverter

Reply:
> No I mean, ConfigEntryError, since we connect to the right IP, but the serial changed, which means the current config entry can't be used with that inverter

Not sure why @bdraco choose to use ConfigEntryNotReady during the overhaul. And wouldn't the net effect be the same, the entry wouldn't load and be signalled?

Reply:
Its likely that discovery will fix it shortly and the next time it tries it will be correct